sling/nlp/parser/trainer/transition-generator.cc

```diff
@@ -0,0 +1,457 @@
+}
+
+// Translates 'action' into an attention-resolved version.
+ParserAction Translate(const Handles &attention, const Action &action) {
+  ParserAction output(action.core.type);
+  if (action.core.length != 0) output.length = action.core.length;
+  if (!action.core.role.IsNil()) output.role = action.core.role;
+
+  switch (output.type) {
+    case ParserAction::EVOKE:
+      output.label = action.frame->type;
+      break;
+    case ParserAction::REFER:
+      output.target = Index(attention, action.frame->handle);
+      break;
+    case ParserAction::EMBED:
+      output.label = action.frame->type;
+      output.target = Index(attention, action.neighbor->handle);
+      break;
+    case ParserAction::ELABORATE:
+      output.label = action.frame->type;
+      output.source = Index(attention, action.neighbor->handle);
+      break;
+    case ParserAction::CONNECT:
+      output.source = Index(attention, action.frame->handle);
+      output.target = Index(attention, action.neighbor->handle);
+      break;
+    case ParserAction::ASSIGN:
+      output.source = Index(attention, action.frame->handle);
+      output.label = action.core.label;
+      break;
+    default:
+    break;
+  }
+
+  return output;
+}
+
+// Updates 'attention' as a result of execution 'action'.
+void Update(const Action &action, Handles *attention) {
+  auto type = action.core.type;
+  if (type == ParserAction::EVOKE || type == ParserAction::EMBED ||
+      type == ParserAction::ELABORATE) {
+    attention->insert(attention->begin(), action.frame->handle);
+  } else if (type == ParserAction::REFER || type == ParserAction::ASSIGN ||
+             type == ParserAction::CONNECT) {
+    std::remove(attention->begin(), attention->end(), action.frame->handle);
+    attention->insert(attention->begin(), action.frame->handle);
+  }
+}
+
+void InitInfo(
+    Frame frame, FrameInfo *frame_info, HandleSet *initialized) {
+  Handle handle = frame.handle();
+  if (initialized->find(handle) != initialized->end()) return;
+  initialized->insert(handle);
+
+  auto *entry = frame_info->Get(handle);
+  std::vector<Frame> pending;
+  for (const Slot &slot : frame) {
+    if (slot.name.IsId()) continue;
+
+    Frame value(frame.store(), slot.value);
+    if (!value.IsFrame()) continue;
+
+    if (slot.name.IsIsA() && entry->type.IsNil()) {
+      entry->type = slot.value;
+    } else {
+      auto *edge = new FrameInfo::Edge(slot.name, false, slot.value);
+      entry->edges.emplace_back(edge);
+      if (slot.value == handle) {
+        edge->inverse = edge;
+        continue;
+      }
+      if (value.IsLocal()) {
+        auto *neighbor = frame_info->Get(slot.value);
+        auto *reverse = new FrameInfo::Edge(slot.name, true, handle);
+        neighbor->edges.emplace_back(reverse);
+        reverse->inverse = edge;
+        edge->inverse = reverse;
+        pending.emplace_back(value);
+      }
+    }
+  }
+
+  for (Frame f : pending) {
+    InitInfo(f, frame_info, initialized);
```

Comment:
What is the reason for not just calling InitInfo recursively inside the loop above?

Reply:
Done. It does cause minor changes in the calling order though, so there are slight differences in the list of output transitions. But I have manually verified and all those diffs are legit.